### PR TITLE
feat: twitter bot setup

### PR DESCRIPTION
**Pull request by Akirtovskis**
Twitter bot that tweets daily the data from last days STX stats

- twitter api interactions using [Twit](https://www.npmjs.com/package/twit )
- cron using scheduling [Bull](https://www.npmjs.com/package/bullmq)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 stacks-node/persistent-data
 node_modules
 data.json
+.env
```
Comment (pradel) [inline .gitignore:4]: AS we are removing the .env let's create a .env.example so users know what to add in local

File: api/jobs/twitterBot.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import { microToStacks, readData } from "../src/utils";
+import { tweet } from "../src/twitterBot/twit";
+
+const currentData = readData();
+
+if (currentData) {
+  const { nbTxsPerDay, uniqueAddressGrowingPerDay, txsFeePerDay } = currentData;
+  tweet(
+    // Date of the day before ( to make sure we have all the data )
+    `Stats from stacks 2.0 blockchain on ${format(
```
Comment (pradel) [inline api/jobs/twitterBot.ts:11]:
```suggestion
    `Stats from Stacks 2.0 blockchain on ${format(
```

File: api/src/twitterBot/bree.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import Bree from "bree";
+
+//This is helper necessary for bree to work w typescript
+function typescript_worker() {
```
Comment (pradel) [inline api/src/twitterBot/bree.ts:4]: this will work only in dev mode, in prod the code is compiled to js so you probably ned to run the worked with ts-node only in dev and make sure that for prod everything is working (run yarn build + yarn start for prod mode)

File: api/src/twitterBot/bree.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  require("ts-node").register();
+  const workerData = require("worker_threads").workerData;
+  require(path.resolve(__dirname, workerData.__filename));
+}
+
+export const bree = new Bree({
+  jobs: [
+    {
+      name: "twitterBot",
+      cron: "0 22 * * *",
```
Comment (pradel) [inline api/src/twitterBot/bree.ts:15]: what time of hour day will this post?

File: api/src/twitterBot/twit.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  access_token: process.env.ACCESS_TOKEN,
+  access_token_secret: process.env.ACCESS_TOKEN_SECRET,
+  timeout_ms: 60 * 1000, // optional HTTP request timeout to apply to all requests.
+  strictSSL: true, // optional - requires SSL certificates to be valid.
+});
+
+export const tweet = (status: string) => {
+  T.post("statuses/update", { status }, function (err, data, response) {
+    console.log(err);
+    console.log(data);
```
Comment (pradel) [inline api/src/twitterBot/twit.ts:15]: let's maybe just print the twitter url to limit the logs emitted?

File: api/src/twitterBot/bree.ts (hunk 4 above)
Comment (Akirtovskis) [reply api/src/twitterBot/bree.ts:15]: @pradel 10PM

File: .gitignore (hunk 1 above)
Comment (Akirtovskis) [reply .gitignore:4]: @pradel added

File: api/src/index.ts [hunk 6]
```diff
@@ -1,13 +1,24 @@
...
 
 let cacheData: FileData | false = false;
 async function generateDataStats() {
   const currentData = readData();
+  if (currentData) {
+    await tweetStats
+      .add(
+        "tweet-stats",
+        { currentData: currentData },
+        { repeat: { cron: "* * * * *" } }
```
Comment (Akirtovskis) [inline api/src/index.ts:18]: @pradel I am facing a problem with the `repeat` part. 

it is adding jobs to queue in some strange way and also not executing them, as you can see here the end time is somewhere very far in the future Screenshot 2021-08-12 at 13 47 32

If I remove the repeat part it works smoothly and tweets immediately. 

I also tried to add intervals, but the same thing happens with that approach too.

File: api/src/index.ts [hunk 7]
```diff
@@ -1,39 +1,58 @@
...
   cacheData = fileData;
 }
 
 generateDataStats()
-  .then(() => {
+  .then(async () => {
     console.log("First data generated");
+
+    ({
+      work: await tweetStatsQueue.removeRepeatableByKey(
```
Comment (pradel) [inline api/src/index.ts:39]: @Akirtovskis is there a way to delete all of them directly?

Comment (pradel) [reply api/src/index.ts:39]: Otherwise if we change the hour at some point it will not delete the old hour

File: api/src/index.ts [hunk 8]
```diff
@@ -1,39 +1,58 @@
...
+        `tweet-stats::::0 22 * * *`
+      ),
+    });
+
+    console.log({ work: await tweetStatsQueue.getRepeatableJobs() });
```
Comment (pradel) [inline api/src/index.ts:44]: let's cleanup all the console.log

File: api/src/twitterBot/bullQueue.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  connection: redisClient,
+});
+export const tweetStatsQueue = new Queue<{}>(queueName, {
+  connection: redisClient,
+});
+
+const worker = new Worker(
+  queueName,
+  async (job) => {
+    const { currentData } = job.data;
```
Comment (pradel) [inline api/src/twitterBot/bullQueue.ts:23]: I removed the currentData when I pushed, it should be read directly from the file :)

File: api/src/twitterBot/twit.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+  access_token: process.env.ACCESS_TOKEN,
+  access_token_secret: process.env.ACCESS_TOKEN_SECRET,
+  timeout_ms: 60 * 1000, // optional HTTP request timeout to apply to all requests.
+  strictSSL: true, // optional - requires SSL certificates to be valid.
+});
+
+export const tweet = (status: string) => {
+  T.post("statuses/update", { status }, function (err, data, response) {
+    console.log(err);
+    //@ts-ignore
```
Comment (pradel) [inline api/src/twitterBot/twit.ts:15]: can we print the link of the tweet and remove the ts-ignore?

File: docker-compose.yml [hunk 11]
```diff
@@ -53,17 +53,25 @@ services:
     networks:
       - stacks-blockchain
 
+  redis:
```
Comment (pradel) [inline docker-compose.yml:56]: We need to apply the same thing to the dev docker compose

File: docker-compose.yml [hunk 12]
```diff
@@ -53,17 +53,25 @@ services:
...
+    # TODO protect via password
```
Comment (pradel) [inline docker-compose.yml:57]: we can do that TODO here or in a separate pr

File: api/src/index.ts (hunk 7 above)
Comment (Akirtovskis) [reply api/src/index.ts:39]: Didn't find a real way to remove all of the directly, so would suggest that for now we go with this. In case of change a simple fix is just to temporary add one more `removeRepeatableByKey` fn with the old values.

File: docker-compose.yml (hunk 12 above)
Comment (Akirtovskis) [reply docker-compose.yml:57]: Let's leave this separate, already created issue : https://github.com/sigle/stxstats/issues/33

File: api/src/twitterBot/bullQueue.ts [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+const redisClient = new Redis();
+
+export const tweetStatsQueueScheduler = new QueueScheduler(queueName, {
+  connection: redisClient,
+});
+export const tweetStatsQueue = new Queue<{}>(queueName, {
+  connection: redisClient,
+});
+
+const currentData = readData();
```
Comment (pradel) [inline api/src/twitterBot/bullQueue.ts:19]: this should be in the worker job to get the latest data when the job is triggered
